File: code_examples/4-Functions/6-By_Value_or_By_Reference/Java/Main.java
/*
 * This program uses a function by reference
 *
 * @author  Mr Coxall
 * @version 1.0
 * @since   2020-09-01
 */

import java.util.Scanner;

public class Main {

  /**
   * This function adds one to a number, by reference.
   * @param someNumber
   */
  public static void addOne(int[] someNumber) {
    someNumber[0] = someNumber[0] + 1;
  }

  public static void main(String[] args) {
    // this function uses a function by reference

    // In Java, you can pass a variable by reference by using an array
    int[] someVariable = new int[1];

    // input
    Scanner input = new Scanner(System.in);
    System.out.print("Enter a number: ");
    someVariable[0] = input.nextInt();

    // process
    addOne(someVariable);

    // output
    System.out.println("The number plus one is: " + someVariable[0]);

    System.out.println("\nDone.");
  }
}

File: code_examples/4-Functions/6-By_Value_or_By_Reference/Python/main.py
#!/usr/bin/env python3
"""
Created by: Mr. Coxall
Created on: Sep 2020
This module uses a function by reference
"""


def add_one(someVariable: List[int]) -> None:
    """The add_one() function adds 1, by reference, returns str."""
    # in python the only way to pass by reference is to pass the whole list

    someVariable[0] = someVariable[0] + 1


def main() -> None:
    """The main() this function gets a number and calls the add_one function, returns None."""
    someNumber = []

    # input
    temp_var = int(input("Enter a number: "))
    someNumber.append(temp_var)

    # process
    add_one(someNumber)

    # output
    print(f"The number plus one is: {someNumber[0]}")

    print("\nDone.")


if __name__ == "__main__":
    main()
